cmd/osm-controller/osm-controller.go

```diff
@@ -210,23 +210,48 @@ func main() {
 	// TODO(draychev): figure out the NS and POD
 	metricsStore := metricsstore.NewMetricStore("TBD_NameSpace", "TBD_PodName")
 
-	// Expose /debug endpoints and data only if the enableDebugServer flag is enabled
-	var debugServer debugger.DebugServer
-	if cfg.IsDebugServerEnabled() {
-		debugServer = debugger.NewDebugServer(certDebugger, xdsServer, meshCatalog, kubeConfig, kubeClient, cfg, kubernetesClient)
-	}
-
 	funcProbes := []health.Probes{xdsServer}
 	httpProbes := getHTTPHealthProbes()
-	httpServer := httpserver.NewHTTPServer(funcProbes, httpProbes, metricsStore, constants.MetricsServerPort, debugServer)
+	httpServer := httpserver.NewHTTPServer(funcProbes, httpProbes, metricsStore, constants.MetricsServerPort)
 	httpServer.Start()
 
-	// Wait for exit handler signal
-	<-stop
+	// Expose /debug endpoints and data only if the enableDebugServer flag is enabled
+	debugServerRunning := !cfg.IsDebugServerEnabled()
+	debugImpl := debugger.NewDebugImpl(certDebugger, xdsServer, meshCatalog, kubeConfig, kubeClient, cfg, kubernetesClient)
+	debugServer := httpserver.NewDebugHTTPServer(debugImpl, constants.DebugPort)
+
+	errCh := make(chan error)
+	go configureDebugServer(debugServer, debugImpl, debugServerRunning, cfg, errCh)
+
+	select {
+	case err := <-errCh:
+		log.Error().Err(err).Msg("Unable to configure debug server")
+
+		// Wait for exit handler signal
+	case <-stop:
+	}
 
 	log.Info().Msg("Goodbye!")
 }
 
+func configureDebugServer(debugServer *httpserver.DebugServer, debugImpl debugger.DebugServer, debugServerRunning bool, cfg configurator.Configurator, errCh chan error) {
+	for range cfg.GetAnnouncementsChannel() {
+		if debugServerRunning && !cfg.IsDebugServerEnabled() {
+			debugServerRunning = false
+			err := debugServer.Stop()
+			if err != nil {
+				log.Error().Err(err).Msg("Unable to stop debug server")
+				errCh <- err
+			}
+		}
+		if !debugServerRunning && cfg.IsDebugServerEnabled() {
+			debugServerRunning = true
+			debugServer = httpserver.NewDebugHTTPServer(debugImpl, constants.DebugPort)
```

Comment:
Since you're already passing in `debugServer`, can you use that instead of creating a new one here?

Reply:
When `debugServer.Stop()` is called it'll call `server.Shutdown()` so the server won't be able to be re-used so then we have to create a new one.